### ISSUE TITLE
Review ScrollingChartNode

### ISSUE THREAD

**Issue by samreid** [phetsims/griddle#30]
From https://github.com/phetsims/wave-interference/issues/24 ScrollingChartNode was introduced for the wave detector tool.  Here is its description:

```js
/**
 * The scrolling graph component.  Like a seismograph, it has pens on the right hand side that record data, and the
 * paper scrolls to the left.  It is currently sized accordingly to be used in a small draggable sensor, like the ones
 * in Wave Interference, Bending Light or Circuit Construction Kit: AC.
 *
 * Please see the demo in http://localhost/griddle/griddle_en.html
 *
 * Moved from wave-interference repo to griddle repo on Wed, Aug 29, 2018.
 *
 * @author Sam Reid (PhET Interactive Simulations)
 */
```

Here's the picture from the griddle demo:

![image](https://user-images.githubusercontent.com/679486/45819785-6df07880-bca2-11e8-8ea9-f45fcb25b7c2.png)

Here's a picture of its usage in Wave Interference:

![image](https://user-images.githubusercontent.com/679486/45819805-7d6fc180-bca2-11e8-9b48-fa351c3aa0e7.png)

Note that the dimensions and gridlines counts are not fully general--this is designed to be a small, probably draggable, sensor.  I don't know if that will require further generalization before 1.0 of wave interference is published.

@ariel-phet can you please select a reviewer for this component?  It is 226 lines of code for ScrollingChartNode plus 40 lines for the demo usage.

**Comment by samreid** [comment]
Questions for the reviewer:

- [x] Should some of the required constructor args be moved to options?
- [x] How would you rename things to avoid having `series.series`?  See line 212.

UPDATE: I addressed these before the review.

**Comment by samreid** [comment]
I noted stringTest=long is not doing so well, self-assigning until that's addressed:

![image](https://user-images.githubusercontent.com/679486/45820206-8f059900-bca3-11e8-816b-1390b38d9879.png)

**Comment by samreid** [comment]
Fixed, but leaving self-assigned to check if there's anything else to do before starting review.

**Comment by samreid** [comment]
I decided to factor out ScrollingChartNode from LabeledChartNode to simplify maintainability and improve customizability in the future.  The files that need review are:
ScrollingChartNode (176 lines)
LabeledScrollingChartNode (121 lines)
GriddleDemoScreenView.demoLabeledScrollingChartNode (48 lines)

I noted the number of lines to quantify the scope/size of what is up for review.  It would be good to have this done (a) before it is used in other simulations such as Bending Light and (b) before Wave Interference goes to RC.  This could be done as part of the Wave Interference code review.  @ariel-phet can you please delegate accordingly?

**Comment by jbphet** [comment]
The code looks organized and well documented.  I particularly like the use of the emitter to trigger internal disposal actions - I'll have to try that.

Here are my comments:

+ My main observation is that it says in the header comment, "Like a seismograph, it has pens on the right hand side that record data, and the paper scrolls to the left", but the paper doesn't scroll, only the data does.  I find this looks strange and unrealistic.  In both the griddle demo and in the Wave Interference sim, it looks to me like the pen on the right side is riding up and down on something, much like in the Wave on a String sim when the "Loose End" option is selected.  I looked for videos of seismographs for reference, and there is a good one at https://www.youtube.com/watch?v=MoOTr70ZYXw.  Note how the background graph either moves or isn't present at all.  If this was an intentional choice due to performance concerns, a canvas node should be considered.
+ There are a number of things that seem like they should be available as options but aren't, such as:
    + background color
    + graph lines on/off
    + graph line options
    + clamp on/off (when off, the data just go off the page, which I had to do in Gene Expression 3rd screen)
+ The `timeProperty` parameter has no units, which may have been intentional to allow this to be very general, but later in the code there is a variable called `maxSeconds`, so I think it should be made clear whether time is assumed to be in seconds or can be any time unit
+ Consider having a type for the `seriesArray` data parameter.  I don't feel super strongly about this, but it seems like it would be useful in future griddle nodes.
+ There are two related variables called `pathNode` and `pathShape` in the `seriesListener` closure.  I would find it more readable if the names were less generic, perhaps something like `dataPathNode`, `dataSeriesNode`, or the like.

**Comment by samreid** [comment]
> My main observation is that it says in the header comment, "Like a seismograph, it has pens on the right hand side that record data, and the paper scrolls to the left", but the paper doesn't scroll, only the data does. I find this looks strange and unrealistic.

The decision to have the data curves move but the grid lines remain fixed was by the design team, including @ariel-phet, @arouinfar @KatieWoe and @kathy-phet.  Perhaps I misspoke in comparing it to a seismograph--I think the gridlines are more like an oscilloscope, where they remain fixed.  If I recall correctly, we preferred the static gridlines because the total elapsed time is immaterial in Wave Interference and the static gridlines make it easier to reference time units. (Wave Interference has a 1 second time reference for water, for example).  I'll create an issue to discuss this point, and to see if we expect future sims to have scrolling gridlines too.

UPDATE: I moved the question about scrolling gridlines to a separate issue for discussion with the design team.

**Comment by samreid** [comment]
> clamp on/off (when off, the data just go off the page, which I had to do in Gene Expression 3rd screen)

Can you elaborate on this point?  I tried to get the Gene Expressions data to go off the chart, but couldn't find a combination of parameters to do so.  Do you think it's likely that design teams will want the ScrollingChart data to fall outside of the chart area?  Or are you saying that ScrollingChartNode data can fall outside of its bounds and it shouldn't?

**Comment by samreid** [comment]
I made several commits to address the recommendations from the review.  For the options, I used the nested options strategy described here: https://github.com/phetsims/phet-info/blob/9942436a3e3d783277fce871f4d346b005868dfe/doc/phet-software-design-patterns.md#nesting

@jbphet back to you for re-review and discussion.  I think I addressed everything aside from https://github.com/phetsims/griddle/issues/30#issuecomment-428044996 and https://github.com/phetsims/griddle/issues/31 which are awaiting feedback.

**Comment by jbphet** [comment]
Sorry, I forgot that it was a bit tricky to make the data go off the chart in the Gene Expression Essentials sim.  Below is what it looks like when this occurs (and the way the clipping looks is admittedly not great since part of the stroke goes above the edge of the chart, but it's the idea that the data would go off the chart instead of being clamped that is what I think we may need):

![image](https://user-images.githubusercontent.com/3082831/46683968-cbdff400-cbae-11e8-9ece-8d58ce6bbecf.png)

**Comment by jbphet** [comment]
The rest of the commits look good, though I think there are two more options that would be useful and should be easy to add:
+ backgroundColor
+ rightGraphMargin (currently a constant)

**Comment by samreid** [comment]
The background color may be customized with e.g.: `graphPanelOptions: {fill: 'blue'}`, using the nested options pattern from the phet-software-design-patterns.md.  In the preceding commit, I added more options, usage of DynamicSeries, and clipping, because I don't believe data should ever render outside of the chart area.  Back to @jbphet for feedback.

**Comment by jbphet** [comment]
I tested the background color and it looks great.

I tried testing the clipping area, and I think it's not quite what we want.  For this test, I started with the usage in `GriddleDemoScreenView` and made the data line red and thicker, and then increased the amplitude of the sine wave, and I got this:

![image](https://user-images.githubusercontent.com/3082831/48231802-ab9d8380-e36c-11e8-8d33-1c45291c846c.png)

I commented out the following line in `dynamicSeriesListener`:
```
const clampedValue = Util.clamp( scaledValue, 0, height );
```
...and then got a graph like this:

![image](https://user-images.githubusercontent.com/3082831/48231737-6e38f600-e36c-11e8-921a-6631ab788b1f.png)

In my opinion, the lower one should be the default behavior, since the upper one could be misleading, potentially causing a viewer to think that the data itself was being clamped and not just the graph.  If @samreid thinks the behavior shown in the top view is desirable in some instances, then there should probably be a "clampData" option.

**Comment by jbphet** [comment]
Also, a minor nit: There are a handful of lines in `ScrollingChartNode` and `LabeledScrollingChartNode` that exceed that 120 column recommendation, and I had to scroll a bit to look at them in my IDE.

**Comment by samreid** [comment]
Good catch @jbphet, I meant to remove `clampedValue` but forgot to do so.  Thanks for the review!  Commits are above.  Let me know if there is anything else.

**Comment by jbphet** [comment]
Looks good to me, thanks for creating this.  Closing.
